### DISCUSSION TITLE
Future ideas

### DISCUSSION THREAD

**Discussion by aalmiray** [moditect/layrry#216] [Ideas]
Here's a set of ideas for future features:

1. Current Layrry is used as an application launcher on the CLI. Given an app descriptor Layrry resolves dependencies and launches a Main class. We keep this behavior of course.
2. **Embedded Layrry**. This would allow modular & non-modular apps to enrich their behavior with pluggable extensions for example. Extensions _must_ be modular but the host application doesn't have to be.
3. **Layrry powered archives/uberjars**. These require packing a minimal Layrry launcher/runner alongside all required app dependencies and application descriptor. Layrry can already resolve all local resources so this would be two kinds of launchers:
 - for archive just resolve dependencies on the filesystem.
 - for uberjar it could either resolve from the JAR itself (need custom resource finder/loader) or unpack to temp directory and resolve from filesystem.

For item No. 3 the assumption is that everything is local so there's no need to include ShrinkWrap dependencies. JCommander would also be redundant as the launcher knows where the application descriptor is. Next, the descriptor's format. We currently support YAML and TOML via external dependencies. In order to minimize the launcher's footprint a different format would be needed. It could simply be `.properties`. The trick would be automatic conversion from YAML/TOML to `.properties` during packaging. This of course would be handled by a Layrry helper as well.

**Comment by bmarwell** [comment]
CC @bdemers

**Comment by bmarwell** [comment]
If properties don't work, you might want to consider XML? It's also part of the JDK. Assuming you're not shipping with jlink (which doesn't make such sense with uber jars, imho), the footprint is the same but the format is much clearer.
However, for shipping with jlink, the footprint is a bit higher if jlink is not used by any dependency inside the uber jar.

**Reply by aalmiray** [reply]
Layer configuration looks like this https://github.com/moditect/layrry/blob/master/examples/plugin-example/greeter-runner/src/test/resources/layers.yml or this https://github.com/moditect/layrry/blob/master/examples/plugin-example/greeter-runner/src/test/resources/layers.toml

The TOML version is closer to what `.properties` could look like. This format means no additional dependencies are needed, nor additional modules other than `java.base`.

**Reply by bmarwell** [reply]
Yes I know. It works until you need lists, so I just wanted to write this down just in case :) 
+1 for properties.

**Reply by aalmiray** [reply]
Given the constrained options of the DSL we could simply use comma separated values when lists are required, such as

```
layers = layrry-platform,api,plugins
layers.layrry-platform.modules = org.moditect.layrry:layrry-platform:1.1.0-SNAPSHOT
layers.api.modules = org.moditect.layrry.example.greeter:greeter-core:1.0.0
layers.api.parents = layrry-platform
layers.plugins.parents = api
layers.plugins.directory = ../../../target/layers
main.module = org.moditect.layrry.example.greeter.core
main.class = org.moditect.layrry.example.greeter.app.App
```
